File: api/app/services/query_classifier.py
"""Query 9 라벨 룰 분류 — production 측 표준 위치.

원래 `evals/auto_goldenset.py` 에 있던 룰 기반 분류 함수를 production 모듈로 이전.
역방향 의존 (`api/app/routers/admin.py` → `evals/auto_goldenset.py` lazy import) 해소.

라벨 9종 (PRD §1.5 + master plan §8.2):
    exact_fact / fuzzy_memory / synonym_mismatch / numeric_lookup /
    table_lookup / vision_diagram / summary / cross_doc / out_of_scope

`/admin/queries/stats` 가 query_type 분포 집계 시 호출. evals 측은 본 모듈을 import
하여 backward-compat 유지 (`from auto_goldenset import classify_query_type` 호출도
`auto_goldenset.py` 가 본 모듈에서 re-export 하므로 그대로 작동).

CLAUDE.md 정합 — DB 호출 0, 외부 의존 0 (stdlib re + typing).
"""

from __future__ import annotations

import re
from typing import Literal

QueryType = Literal[
    "exact_fact",
    "fuzzy_memory",
    "synonym_mismatch",
    "numeric_lookup",
    "table_lookup",
    "vision_diagram",
    "summary",
    "cross_doc",
    "out_of_scope",
]

QUERY_TYPE_LABELS: tuple[QueryType, ...] = (
    "exact_fact",
    "fuzzy_memory",
    "synonym_mismatch",
    "numeric_lookup",
    "table_lookup",
    "vision_diagram",
    "summary",
    "cross_doc",
    "out_of_scope",
)

# 룰 키워드 (substring 매칭 — 한국어 토큰화 의존 0)
_VISION_KEYWORDS: tuple[str, ...] = (
    "다이어그램", "그림", "도식", "구조도", "이미지", "사진", "도표",
)
_TABLE_KEYWORDS: tuple[str, ...] = (
    "표", "리스트", "목록", "별표", "카테고리", "항목 목록",
)
_SUMMARY_KEYWORDS: tuple[str, ...] = (
    "요약", "핵심", "정리", "개요", "짧게", "한줄", "한 줄",
)
_CROSS_DOC_KEYWORDS: tuple[str, ...] = (
    "비교", "차이", "대비", "달라", "차이점",
)
_FUZZY_KEYWORDS: tuple[str, ...] = (
    "그때", "어디 있더라", "어디 있었", "뭐였지", "있었나", "있었지",
    "었더라", "았더라", "기억나",
)
_NUMERIC_PATTERNS: tuple[re.Pattern[str], ...] = (
    # 단위 alternation 은 **긴 것 먼저** — "개월" 이 "개" 보다 먼저 매칭되도록.
    re.compile(r"\d+(?:\.\d+)?\s*(?:개월|시간|kg|km|cm|%|원|년|월|일|회|건|개|점|명|분|초|m)"),
    re.compile(r"몇\s*[가-힣]"),
    re.compile(r"얼마"),
)
_NUMERIC_KEYWORDS: tuple[str, ...] = (
    "얼마", "금액", "가격", "비용", "수치", "수량", "개수", "지원금",
)

# 동의어 쌍 — query 가 한쪽 표현, source 가 반대편 표현이면 synonym_mismatch.
# (a, b): query 안에 a 가 있고 source 안에 b 가 있거나 반대일 때.
_SYNONYM_PAIRS: tuple[tuple[str, str], ...] = (
    ("개인정보", "비식별화"),
    ("환자 정보", "비식별화"),
    ("색상", "컬러"),
    ("시트", "가죽"),
    ("규정", "내규"),
    ("직원", "임직원"),
    ("회의", "협의"),
)


def classify_query_type(
    query: str,
    *,
    source_chunk_text: str = "",
    expected_doc_titles: list[str] | None = None,
    is_negative: bool = False,
) -> QueryType:
    """query → 9 라벨 중 1개 (룰 기반).

    우선순위:
    1. is_negative=True → out_of_scope
    2. vision_diagram (그림/다이어그램 키워드)
    3. table_lookup (표/목록 키워드)
    4. cross_doc (비교 키워드 또는 doc_title 2개 이상)
    5. numeric_lookup (숫자 패턴 또는 금액 키워드)
    6. summary (요약 키워드)
    7. synonym_mismatch (동의어 쌍 cross 매칭)
    8. fuzzy_memory (흐릿한 톤 키워드)
    9. exact_fact (default)
    """
    if is_negative:
        return "out_of_scope"

    q = query.strip()

    if any(kw in q for kw in _VISION_KEYWORDS):
        return "vision_diagram"

    if any(kw in q for kw in _TABLE_KEYWORDS):
        return "table_lookup"

    if expected_doc_titles and len(expected_doc_titles) >= 2:
        return "cross_doc"
    if any(kw in q for kw in _CROSS_DOC_KEYWORDS):
        return "cross_doc"

    if any(p.search(q) for p in _NUMERIC_PATTERNS):
        return "numeric_lookup"
    if any(kw in q for kw in _NUMERIC_KEYWORDS):
        return "numeric_lookup"

    if any(kw in q for kw in _SUMMARY_KEYWORDS):
        return "summary"

    if source_chunk_text:
        for term_a, term_b in _SYNONYM_PAIRS:
            in_query_a = term_a in q
            in_query_b = term_b in q
            in_source_a = term_a in source_chunk_text
            in_source_b = term_b in source_chunk_text
            # query 가 한쪽, source 가 반대편이면 synonym_mismatch
            if (in_query_a and in_source_b and not in_source_a) or (
                in_query_b and in_source_a and not in_source_b
            ):
                return "synonym_mismatch"

    if any(kw in q for kw in _FUZZY_KEYWORDS):
        return "fuzzy_memory"

    return "exact_fact"

File: supabase/functions/_shared/query_classifier.ts
/**
 * `api/app/services/query_classifier.py` 포팅 — 질의를 9 라벨로 나누는 룰 분류기.
 *
 * `/admin/queries/stats` 의 `query_type_distribution` 이 이 함수를 질의마다 부른다.
 * DB·외부 호출 0, 순수 함수다.
 *
 * ## Python 정규식을 그대로 옮기면 틀린다
 * | 원본 | Python | JS 기본 |
 * |---|---|---|
 * | `\d` | 유니코드 Nd 전부(680 자) — `"３"`, `"٣"` 도 숫자 | ASCII `0-9` 만 |
 * | `\s` | `\x1c-\x1f`, `\x85` 포함 / `U+FEFF` 제외 | 그 반대 |
 * | `strip()` | 위 공백 집합 | `trim()` 은 `U+FEFF` 를 뗀다 |
 *
 * 그래서 `\d` → `\p{Nd}`, `\s` → `[PY_SP]`, `strip()` → `pyStrip()` 으로 옮기고
 * `u` 플래그를 붙였다. 셋 다 대조에서 실제로 갈린 적이 있는 항목이다.
 *
 * ## `/admin` 에서는 인자 3개가 안 쓰인다
 * 호출부가 `classify_query_type(query)` 뿐이라 `sourceChunkText` ·
 * `expectedDocTitles` · `isNegative` 는 늘 기본값이다 → `synonym_mismatch` 와
 * "제목 2개 이상 → cross_doc" 분기는 **admin 경로에서 도달 불가능**하다.
 * 그래도 원본 그대로 옮겼다. 지금 빼면 나중에 evals 를 옮길 때 조용히 갈린다.
 */

import { PY_SP, pyStrip } from "./search/pystr.ts";

export type QueryType =
  | "exact_fact"
  | "fuzzy_memory"
  | "synonym_mismatch"
  | "numeric_lookup"
  | "table_lookup"
  | "vision_diagram"
  | "summary"
  | "cross_doc"
  | "out_of_scope";

/** 응답 스키마가 **항상 9 키**를 노출한다 — 0 건 라벨도 프론트가 행으로 그린다. 순서 유지. */
export const QUERY_TYPE_LABELS: readonly QueryType[] = [
  "exact_fact",
  "fuzzy_memory",
  "synonym_mismatch",
  "numeric_lookup",
  "table_lookup",
  "vision_diagram",
  "summary",
  "cross_doc",
  "out_of_scope",
];

const VISION_KEYWORDS = ["다이어그램", "그림", "도식", "구조도", "이미지", "사진", "도표"];
const TABLE_KEYWORDS = ["표", "리스트", "목록", "별표", "카테고리", "항목 목록"];
const SUMMARY_KEYWORDS = ["요약", "핵심", "정리", "개요", "짧게", "한줄", "한 줄"];
const CROSS_DOC_KEYWORDS = ["비교", "차이", "대비", "달라", "차이점"];
const FUZZY_KEYWORDS = [
  "그때",
  "어디 있더라",
  "어디 있었",
  "뭐였지",
  "있었나",
  "있었지",
  "었더라",
  "았더라",
  "기억나",
];

// 단위 alternation 은 **긴 것 먼저** — "개월" 이 "개" 보다 먼저 매칭돼야 한다.
// 순서를 바꾸면 "3개월" 이 "3개" + "월" 로 쪼개져도 결과는 같지만, 원본 정규식과
// 다른 물건이 되므로 그대로 둔다.
const NUMERIC_PATTERNS: RegExp[] = [
  new RegExp(
    `\\p{Nd}+(?:\\.\\p{Nd}+)?[${PY_SP}]*(?:개월|시간|kg|km|cm|%|원|년|월|일|회|건|개|점|명|분|초|m)`,
    "u",
  ),
  new RegExp(`몇[${PY_SP}]*[가-힣]`, "u"),
  /얼마/u,
];
const NUMERIC_KEYWORDS = ["얼마", "금액", "가격", "비용", "수치", "수량", "개수", "지원금"];

/** query 가 한쪽 표현, source 가 반대편 표현이면 `synonym_mismatch`. */
const SYNONYM_PAIRS: [string, string][] = [
  ["개인정보", "비식별화"],
  ["환자 정보", "비식별화"],
  ["색상", "컬러"],
  ["시트", "가죽"],
  ["규정", "내규"],
  ["직원", "임직원"],
  ["회의", "협의"],
];

export interface ClassifyOptions {
  sourceChunkText?: string;
  expectedDocTitles?: string[] | null;
  isNegative?: boolean;
}

/**
 * 질의 → 9 라벨 하나. 우선순위는 원본 순서 그대로다.
 *
 * 1 `isNegative` → out_of_scope · 2 vision · 3 table · 4 cross_doc ·
 * 5 numeric · 6 summary · 7 synonym · 8 fuzzy · 9 exact_fact(기본)
 */
export function classifyQueryType(
  query: string,
  opts: ClassifyOptions = {},
): QueryType {
  if (opts.isNegative) return "out_of_scope";

  const q = pyStrip(query);

  if (VISION_KEYWORDS.some((kw) => q.includes(kw))) return "vision_diagram";
  if (TABLE_KEYWORDS.some((kw) => q.includes(kw))) return "table_lookup";

  const titles = opts.expectedDocTitles;
  if (titles && titles.length >= 2) return "cross_doc";
  if (CROSS_DOC_KEYWORDS.some((kw) => q.includes(kw))) return "cross_doc";

  if (NUMERIC_PATTERNS.some((p) => p.test(q))) return "numeric_lookup";
  if (NUMERIC_KEYWORDS.some((kw) => q.includes(kw))) return "numeric_lookup";

  if (SUMMARY_KEYWORDS.some((kw) => q.includes(kw))) return "summary";

  const src = opts.sourceChunkText ?? "";
  if (src) {
    for (const [a, b] of SYNONYM_PAIRS) {
      const inQueryA = q.includes(a);
      const inQueryB = q.includes(b);
      const inSourceA = src.includes(a);
      const inSourceB = src.includes(b);
      if (
        (inQueryA && inSourceB && !inSourceA) ||
        (inQueryB && inSourceA && !inSourceB)
      ) {
        return "synonym_mismatch";
      }
    }
  }

  if (FUZZY_KEYWORDS.some((kw) => q.includes(kw))) return "fuzzy_memory";

  return "exact_fact";
}
